lib/peer.ml

```diff
@@ -24,3 +24,36 @@ let from_sockaddr sockaddr =
 
 let to_sockaddr peer =
   Lwt_unix.ADDR_INET (Unix.inet_addr_of_string peer.address, peer.port)
+
+(* Basic Knuth shuffle => https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle *)
+let knuth_shuffle known_peers =
+  let initial_array = Array.length (Array.of_list known_peers) in
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
+  for i = initial_array - 1 downto 1 do
+    let k = Random.int (i + 1) in
+    let x = shuffled_array.(k) in
+    shuffled_array.(k) <- shuffled_array.(i);
+    shuffled_array.(i) <- x
+  done;
+  Array.to_list shuffled_array
+
+(* Regarding the SWIM protocol, the list of peers is not ordered.
+   Hence, I basically went for a shuffle after adding the new peer *)
+let add_peer peer_to_add peer =
+  let new_peers = peer_to_add :: peer.known_peers in
```

Comment:
Come to think of it, we should probably store peers in a data structure that allows us to look up a peer by address in estimated O(1) time. How about OCaml Map?

Reply:
Totally reworked this function.